File: 06/BinarySearch2/BinarySearch2.c
#include <stdlib.h> 
#include <stdio.h>
#include "Point.h"

int ComparePoint( const void *_elem1, const void *_elem2 ) 
{ 
    Point* elem1 = (Point*)_elem1; 
    Point* elem2 = (Point*)_elem2; 
 
    if ( elem1->point > elem2->point ) 
        return 1; 
    else if ( elem1->point < elem2->point ) 
        return -1; 
    else 
        return 0;     
} 


int main( void )
{
    int Length = sizeof DataSet / sizeof DataSet[0];     
    Point  target;
    Point* found  = NULL;
 
    //  점수의 오름차순으로 정렬 
    qsort( (void*)DataSet, Length, sizeof (Point), ComparePoint );

    //  671.78 점을 받은 학생 찾기 
    target.id     = 0;
    target.point  = 671.78;

    found = bsearch( 
             (void*)&target, 
             (void*)DataSet, 
             Length, 
             sizeof (Point), 
             ComparePoint );
    
    printf("found... ID: %d, Point: %f \n", found->id, found->point );
 
    return 0;
}

File: Rust/ch06/src/bin/binary_search2.rs
// 표준 라이브러리 이진 탐색 (BinarySearch2)
// Clang/06/BinarySearch2 포팅. C의 qsort + bsearch 를 sort_by + binary_search_by 로 옮겼다.
use ch06::points_data::{data_set, Point};
use ch06::search::compare_point;

fn main() {
    let mut data = data_set();

    // 점수의 오름차순으로 정렬
    data.sort_by(compare_point);

    // 671.78 점을 받은 학생 찾기
    let target = Point { id: 0, point: 671.78 };

    let found = data
        .binary_search_by(|probe| compare_point(probe, &target))
        .ok()
        .map(|idx| &data[idx])
        .unwrap();

    println!("found... ID: {}, Point: {:.6} ", found.id, found.point);
}
